diff --git a/platform/fpga/ariane/platform.c b/platform/fpga/ariane/platform.c
index c6f0ffd..2408514 100644
--- a/platform/fpga/ariane/platform.c
+++ b/platform/fpga/ariane/platform.c
@@ -24,10 +24,10 @@
 #define ARIANE_UART_REG_SHIFT			2
 #define ARIANE_UART_REG_WIDTH			4
 #define ARIANE_PLIC_ADDR			0xc000000
-#define ARIANE_PLIC_NUM_SOURCES			3
+#define ARIANE_PLIC_NUM_SOURCES			32
 #define ARIANE_HART_COUNT			1
 #define ARIANE_CLINT_ADDR			0x2000000
-#define ARIANE_ACLINT_MTIMER_FREQ		1000000
+#define ARIANE_ACLINT_MTIMER_FREQ		25000000
 #define ARIANE_ACLINT_MSWI_ADDR			(ARIANE_CLINT_ADDR + \
 						 CLINT_MSWI_OFFSET)
 #define ARIANE_ACLINT_MTIMER_ADDR		(ARIANE_CLINT_ADDR + \
