diff --git a/drivers/net/xilinx_axi_emac.c b/drivers/net/xilinx_axi_emac.c
index 2ce6271afe..8f3eb6ee59 100644
--- a/drivers/net/xilinx_axi_emac.c
+++ b/drivers/net/xilinx_axi_emac.c
@@ -75,6 +75,40 @@ DECLARE_GLOBAL_DATA_PTR;
 
 #define DMAALIGN	128
 
+#define Z_STRINGIFY(x) #x
+#define STRINGIFY(s) Z_STRINGIFY(s)
+
+#define RISCV_MISC_MEM_OPCOPE 0b0001111
+#define RISCV_FUNC3_CBO 0b010
+
+#define RISCV_IMM_CBO_FLUSH 0b000000000010
+#define RISCV_IMM_CBO_CLEAN 0b000000000001
+#define RISCV_IMM_CBO_INVAL 0b000000000000
+
+#define CBO_FLUSH_STR(REG) ".insn i " STRINGIFY(RISCV_MISC_MEM_OPCOPE) "," STRINGIFY(RISCV_FUNC3_CBO) ", zero," STRINGIFY(REG) "," STRINGIFY(RISCV_IMM_CBO_FLUSH)
+#define CBO_CLEAN_STR(REG) ".insn i " STRINGIFY(RISCV_MISC_MEM_OPCOPE) "," STRINGIFY(RISCV_FUNC3_CBO) ", zero," STRINGIFY(REG) "," STRINGIFY(RISCV_IMM_CBO_CLEAN)
+#define CBO_INVAL_STR(REG) ".insn i " STRINGIFY(RISCV_MISC_MEM_OPCOPE) "," STRINGIFY(RISCV_FUNC3_CBO) ", zero," STRINGIFY(REG) "," STRINGIFY(RISCV_IMM_CBO_INVAL)
+
+#define CONFIG_DCACHE_LINE_SIZE 16
+
+void flush_dcache_range(unsigned long addr, unsigned long end_addr){
+	const size_t size = end_addr - addr;
+    const size_t num_cachelines_affected = (size + CONFIG_DCACHE_LINE_SIZE-1) / CONFIG_DCACHE_LINE_SIZE;
+    uint8_t *flush_addr = (uint8_t *)addr;
+	size_t cacheline;
+
+    /* wait for any previous stores to commit */
+     __asm__ volatile("fence rw,rw");
+
+    for(cacheline = 0; cacheline < num_cachelines_affected; cacheline++){
+        __asm__ volatile("mv t0, %0\n\t" CBO_FLUSH_STR (t0) "\n\t"::"r"(flush_addr):"t0");
+        flush_addr += CONFIG_DCACHE_LINE_SIZE;
+    }
+
+	/* need to wait for the CMO to commit */
+    __asm__ volatile("fence rw,rw");
+}
+
 static u8 rxframe[PKTSIZE_ALIGN] __attribute((aligned(DMAALIGN)));
 
 /* Reflect dma offsets */
